lib/maxwell/conn.ex

```diff
@@ -128,222 +169,269 @@ defmodule Maxwell.Conn do
+  @spec put_query_string(Conn.t, map()) :: Conn.t | no_return
+  def put_query_string(%Conn{state: :unsent, query_string: qs} = conn, query) do
+    %{conn | query_string: Map.merge(qs, query)}
   end
   def put_query_string(_conn, _query_map), do: raise AlreadySentError
 
-  @doc """
-  Add query string to `conn.query_string`.
+  @doc false
+  def put_query_string(query) when is_map(query) do
+    IO.warn "put_query_string/1 is deprecated, use new/1 or new/2 followed by put_query_string/2 instead"
+    put_query_string(new(), query)
+  end
 
-  * `conn` - `%Conn{}`
-  * `key` - query key, for example `"name"`.
-  * `value` - query value, for example `"lucy"`.
+  @doc """
+  Set a query string value for the request.
 
   ### Examples
 
-        # %Conn{query_string: %{name: "zhong wen"}}
-        put_query_string(%Conn{}, :name, "zhong wen")
-
+        iex> put_query_string(new(), :name, "zhong wen")
+        %Maxwell.Conn{query_string: %{:name => "zhong wen"}}
   """
-  def put_query_string(conn = %Conn{state: :unsent, query_string: query_string}, key, value) do
-    %{conn| query_string: Map.put(query_string, key, value)}
+  def put_query_string(%Conn{state: :unsent, query_string: qs} = conn, key, value) do
+    %{conn | query_string: Map.put(qs, key, value)}
   end
   def put_query_string(_conn, _key, _value), do: raise AlreadySentError
 
   @doc """
-  Merge http headers.
-
-    * `conn` - `%Conn{}`
-    * `req_headers` - reqeust headers, for example `%{"content-type" => "text/javascript"}`
+  Merge a map of headers into the existing headers of the connection.
 
   ### Examples
 
-      # %Conn{headers: %{"Content-Type" => "application/json", "User-Agent" => "zhongwenool"}
-      %Conn{headers: %{"Content-Type" => "text/javascript"}
-      |> put_req_header(%{"Content-Type" => "application/json"})
-      |> put_req_header(%{"User-Agent" => "zhongwencool"})
-
+      iex> %Maxwell.Conn{headers: %{"content-type" => "text/javascript"}
+      |> put_req_headers(%{"Accept" => "application/json"})
+      %Maxwell.Conn{req_headers: %{"accept" => "application/json", "content-type" => "text/javascript"}}
   """
-  def put_req_header(conn \\ %Conn{}, map_headers)
-  def put_req_header(conn = %Conn{state: :unsent, req_headers: headers}, map_headers)when is_map(map_headers) do
-    downcase_func =
-    fn({downcase_header, {_original_header, _header_value} = value}, acc) ->
-      Map.put(acc, downcase_header, value)
-      ({header, _val} = value, acc) ->
-        Map.put(acc, String.downcase(header), value)
-    end
-    new_headers = Enum.reduce(map_headers, headers, downcase_func)
-    %{conn| req_headers: new_headers}
+  @spec put_req_headers(Conn.t, map()) :: Conn.t | no_return
+  def put_req_headers(%Conn{state: :unsent, req_headers: headers} = conn, extra_headers) when is_map(extra_headers) do
+    new_headers =
+    extra_headers
+    |> Enum.reduce(headers, fn {header_name, header_value}, acc ->
+      acc
+      |> Map.delete(header_name)
```

Comment:
why need delete header_name first?
It's seems to have guaranteed that the head is all lowercase.